File: 数据结构/有效的括号.py

class Solution:
    def isValid(self, s: str) -> bool:
        mp = {'}':'{', ']':'[', ')':'('}
        stk = []
        for i, c in enumerate(s):
            if c in "{[(":
                stk.append(c)
            else: # c in "}])"
                if len(stk) != 0 and mp[c] == stk[-1]:
                    stk.pop()
                else:
                    return False
        
        return len(stk) == 0

File: 100-学习/110-计算机/数据结构/有效的括号.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    bool isValid(string s) {
        unordered_map<char, char> mp = {
                {')', '('},
                {']', '['},
                {'}', '{'}};

        string leftBrackets = "([{";// 只创建一次
        stack<char> stk;

        for (char c: s) {
            if (leftBrackets.find(c) != string::npos) {
                stk.push(c);
            } else {
                if (stk.empty() || mp[c] != stk.top()) {
                    return false;
                }
                stk.pop();
            }
        }

        return stk.empty();
    }
};